### ISSUE TITLE
Adjust our defaults for Vpc, and provide advanced configurability for a Vpc.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [pulumi/pulumi-awsx#350]
The existing Vpc has some problems with its approach in this space.  Specifically:

1. Our usage of all addresses in the vpc cidr block violates [best practices](https://aws.amazon.com/answers/networking/aws-single-vpc-design/): "Do not allocate all network addresses at once; instead ensure that you reserve some address space for future use."
2.  Once created, it is virtually impossible to change an existing Vpc.

To that end, we propose a few changes to Vpc that will likely need to come in a major package upgrade (i.e. a v0.19 release of awsx).

1. We propose changing the defaults of `new Vpc("no")` (i.e. without args) to literally produce a Vpc as per https://docs.aws.amazon.com/vpc/latest/userguide/default-vpc.html .  This means a `/20` subnet size, using all availability zones, and only creating `public` subnets.   This is a change from our previous behavior of using 2 availability zones, having `public` and `private` subnets, and having those subnets consume the entirety of the space available.

2. We propose a option on `new Vpc` to ask for the old defaults.  This will be necessary to not impact existing stacks that use the defaults.  However, we think it's ok for that to be opt-in behavior.  i.e. to optimize for new programs versus catering to existing ones.

3. We propose that on top of the simple way to create Vpcs there be an arg to them which is a list of the actual `realized` subnet-descriptors that allow the callee to be very explicit about subnet layout.  

4. We propose that Vpcs have a way of dumping out this `realized subnet descriptors` information.  This provides a 'graduate up' path where people can take an existing simple Vpc, and get the subnet-descriptor information so thaat they can then add more subnets without affecting existing ones.

For example, if a user started with `new Vpc` they could then dumb the descriptors getting back something like this `[{ name: "subnet1", type: "public", cidrBlock: ".../20" }, { name: "subnet2", type: "public", cidrBlock: ".../20" }]`  They could then change this array directly, leaving in any subnets they didn't want to change, and adding/removing/changing anything else they cared about.

Subnet descriptors would likely be similar to something like this:

```ts
interface SubnetDescription {
    type: x.ec2.VpcSubnetType;
    subnetName: string;
    availabilityZone: number;
    cidrBlock: string;
    tags?: pulumi.Input<aws.Tags>;
    mapPublicIpOnLaunch?: pulumi.Input<boolean>;
    assignIpv6AddressOnCreation?: pulumi.Input<boolean>;
}
```

**Comment by CyrusNajmabadi** [comment]
Tagging @lukehoban @clstokes

**Comment by lukehoban** [comment]
For significant changes like this, let's hold for a 1.1 or 2.0 release.

**Comment by aavileli** [comment]
Yes. please fix this allocates full address space without providing any overrides

**Comment by jandom** [comment]
Yes, all the way yes – this is a silent killer, I nearly didn't realized it. Played around with the tool, added some isolated subnets and created the initial VPC. The added a new subnet and expected a magical re-adjustment to happen, however I got a 400 saying that CIDR blocks overlap.

**Comment by seivan** [comment]
What's the status on this ticket?

**Comment by jkodroff** [comment]
Closing: Addressed in AWSX 1.x.
